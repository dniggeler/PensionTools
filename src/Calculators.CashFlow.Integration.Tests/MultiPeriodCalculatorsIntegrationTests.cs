using System;
using System.Net.Http;
using System.Net.Http.Json;
using System.Threading.Tasks;
using Domain.Enums;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Mvc.Testing;
using Microsoft.AspNetCore.TestHost;
using Microsoft.Extensions.Configuration;
using PensionCoach.Tools.CommonTypes;
using PensionCoach.Tools.CommonTypes.MultiPeriod;
using PensionCoach.Tools.CommonTypes.MultiPeriod.Definitions;
using Snapshooter.Xunit;
using TaxCalculator.WebApi;
using Xunit;

namespace Calculators.CashFlow.Integration.Tests;

[Trait("Cash-Flow", "Integration")]
public class MultiPeriodCalculatorsIntegrationTests : IClassFixture<WebApplicationFactory<Startup>>
{
    private readonly HttpClient client;

    public MultiPeriodCalculatorsIntegrationTests()
    {
        var testServer = new TestServer(
            new WebHostBuilder()
                .ConfigureAppConfiguration((_, builder) =>
                {
                    builder.AddJsonFile("appsettings.integration.json");
                })
                .UseStartup<Startup>());

        client = testServer.CreateClient();
        client.BaseAddress = new Uri("http://localhost/api/calculator/");
    }

    [Fact(DisplayName = "Wealth Only")]
    public async Task Calculate_Wealth_Only()
    {
        MultiPeriodRequest request = GetRequest();

        var response = await client.PostAsJsonAsync("multiperiod", request);

        response.EnsureSuccessStatusCode();

        var result = await response.Content.ReadAsStringAsync();

        Snapshot.Match(result);
    }

    [Fact(DisplayName = "Wealth And Income")]
    public async Task Calculate_Wealth_And_Income()
    {
        // given
        decimal income = 100000;

        MultiPeriodRequest request = GetRequest();

        request.CashFlowDefinitionRequest.SalaryPaymentsDefinition = new SalaryPaymentsDefinition
        {
            YearlyAmount = income,
            DateOfEndOfPeriod = new DateTime(2034, 4, 1),
        };

        var response = await client.PostAsJsonAsync("multiperiod", request);

        response.EnsureSuccessStatusCode();

        var result = await response.Content.ReadAsStringAsync();

        Snapshot.Match(result);
    }

    private static MultiPeriodRequest GetRequest()
    {
        decimal wealth = 500000;

        return new MultiPeriodRequest
        {
            StartingYear = 2023,
            NumberOfPeriods = 10,
            Name = "Test Multi-Period Calculator",
            BfsMunicipalityId = 261,
            CivilStatus = CivilStatus.Single,
            Income = 0,
            Wealth = wealth,
            DateOfBirth = new DateTime(1980, 1, 1),
            ReligiousGroupType = ReligiousGroupType.Other,
            Gender = Gender.Male,
            CashFlowDefinitionRequest = new()
            {
                SetupAccountDefinition = new SetupAccountDefinition
                {
                    InitialWealth = wealth
                }
            }
        };
    }
}
