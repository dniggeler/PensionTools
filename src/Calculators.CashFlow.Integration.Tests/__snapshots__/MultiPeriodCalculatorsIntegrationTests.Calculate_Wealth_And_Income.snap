{"startingYear":2023,"numberOfPeriods":12,"accounts":[{"year":2023,"amount":-530000.00,"accountType":"Exogenous"},{"year":2023,"amount":0,"accountType":"Income"},{"year":2023,"amount":512574.6350,"accountType":"Wealth"},{"year":2023,"amount":0,"accountType":"OccupationalPension"},{"year":2023,"amount":0,"accountType":"ThirdPillar"},{"year":2023,"amount":17425.3650,"accountType":"Tax"},{"year":2024,"amount":-560000.00,"accountType":"Exogenous"},{"year":2024,"amount":0,"accountType":"Income"},{"year":2024,"amount":525122.990000,"accountType":"Wealth"},{"year":2024,"amount":0,"accountType":"OccupationalPension"},{"year":2024,"amount":0,"accountType":"ThirdPillar"},{"year":2024,"amount":34877.010000,"accountType":"Tax"},{"year":2025,"amount":-590000.00,"accountType":"Exogenous"},{"year":2025,"amount":0,"accountType":"Income"},{"year":2025,"amount":537642.87500000,"accountType":"Wealth"},{"year":2025,"amount":0,"accountType":"OccupationalPension"},{"year":2025,"amount":0,"accountType":"ThirdPillar"},{"year":2025,"amount":52357.12500000,"accountType":"Tax"},{"year":2026,"amount":-620000.00,"accountType":"Exogenous"},{"year":2026,"amount":0,"accountType":"Income"},{"year":2026,"amount":550136.4800000000,"accountType":"Wealth"},{"year":2026,"amount":0,"accountType":"OccupationalPension"},{"year":2026,"amount":0,"accountType":"ThirdPillar"},{"year":2026,"amount":69863.5200000000,"accountType":"Tax"},{"year":2027,"amount":-650000.00,"accountType":"Exogenous"},{"year":2027,"amount":0,"accountType":"Income"},{"year":2027,"amount":562601.615000000000,"accountType":"Wealth"},{"year":2027,"amount":0,"accountType":"OccupationalPension"},{"year":2027,"amount":0,"accountType":"ThirdPillar"},{"year":2027,"amount":87398.385000000000,"accountType":"Tax"},{"year":2028,"amount":-680000.00,"accountType":"Exogenous"},{"year":2028,"amount":0,"accountType":"Income"},{"year":2028,"amount":575040.47000000000000,"accountType":"Wealth"},{"year":2028,"amount":0,"accountType":"OccupationalPension"},{"year":2028,"amount":0,"accountType":"ThirdPillar"},{"year":2028,"amount":104959.53000000000000,"accountType":"Tax"},{"year":2029,"amount":-710000.00,"accountType":"Exogenous"},{"year":2029,"amount":0,"accountType":"Income"},{"year":2029,"amount":587450.8550000000000000,"accountType":"Wealth"},{"year":2029,"amount":0,"accountType":"OccupationalPension"},{"year":2029,"amount":0,"accountType":"ThirdPillar"},{"year":2029,"amount":122549.1450000000000000,"accountType":"Tax"},{"year":2030,"amount":-740000.00,"accountType":"Exogenous"},{"year":2030,"amount":0,"accountType":"Income"},{"year":2030,"amount":599834.960000000000000000,"accountType":"Wealth"},{"year":2030,"amount":0,"accountType":"OccupationalPension"},{"year":2030,"amount":0,"accountType":"ThirdPillar"},{"year":2030,"amount":140165.040000000000000000,"accountType":"Tax"},{"year":2031,"amount":-770000.00,"accountType":"Exogenous"},{"year":2031,"amount":0,"accountType":"Income"},{"year":2031,"amount":612192.78500000000000000000,"accountType":"Wealth"},{"year":2031,"amount":0,"accountType":"OccupationalPension"},{"year":2031,"amount":0,"accountType":"ThirdPillar"},{"year":2031,"amount":157807.21500000000000000000,"accountType":"Tax"},{"year":2032,"amount":-800000.00,"accountType":"Exogenous"},{"year":2032,"amount":0,"accountType":"Income"},{"year":2032,"amount":624522.1400000000000000000000,"accountType":"Wealth"},{"year":2032,"amount":0,"accountType":"OccupationalPension"},{"year":2032,"amount":0,"accountType":"ThirdPillar"},{"year":2032,"amount":175477.8600000000000000000000,"accountType":"Tax"},{"year":2033,"amount":-830000.00,"accountType":"Exogenous"},{"year":2033,"amount":0,"accountType":"Income"},{"year":2033,"amount":636825.21500000000000000000000,"accountType":"Wealth"},{"year":2033,"amount":0,"accountType":"OccupationalPension"},{"year":2033,"amount":0,"accountType":"ThirdPillar"},{"year":2033,"amount":193174.78500000000000000000000,"accountType":"Tax"},{"year":2034,"amount":-837500.0000,"accountType":"Exogenous"},{"year":2034,"amount":0.00,"accountType":"Income"},{"year":2034,"amount":641908.82000000000000000000000,"accountType":"Wealth"},{"year":2034,"amount":0,"accountType":"OccupationalPension"},{"year":2034,"amount":0,"accountType":"ThirdPillar"},{"year":2034,"amount":195591.18000000000000000000000,"accountType":"Tax"}]}
